### PR TITLE
convert batchnorm with synncbatchnorm

### PR DESCRIPTION
**Pull request by yonigottesman**
Fix issue  #269
Hi @owenvallis 
I added a function `convert_sync_batchnorm` which replaces the batchnorm layers with syncbatchnorm **in place**.
The names of the variables are the same same so I was able to load the weights without any problem. To check all the weights are the same (trainable + batchnorm untrainable moving_mean moving std) I ran this successfully:
~~~python
def compare_model(m1,m2):
    for l1,l2 in zip(m1.layers,m2.layers):
        if len(l1.get_weights()) != 0:
            assert len(l1.get_weights()) == len(l2.get_weights())
            for i,j in zip(l1.get_weights(),l2.get_weights()):
                assert tf.reduce_all(i == j)
                
    r = tf.random.uniform([16,224, 224, 3])
    out1 = r1.predict(r)
    out2 = r2.predict(r)
    return tf.reduce_mean(abs(out1 - out2))

r1 = build_resnet(weights="imagenet")
r2 = tf.keras.applications.resnet50.ResNet50(include_top=False,weights="imagenet")
compare_model(r1,r2)

e1 = build_effnet(variant="B0",weights="imagenet")
e2 = tf.keras.applications.EfficientNetB0(include_top=False,weights="imagenet")
compare_model(e1,e2)
~~~

BUT the return values ( the result of passing r through both the models) are not bit exact. the output of r1,r1 is
`<tf.Tensor: shape=(), dtype=float32, numpy=1.9440341e-07>`
and output of e1 e2 is 
`<tf.Tensor: shape=(), dtype=float32, numpy=1.8642845e-07>`

I found the reason of this really small error is that the SyncedBatchNormalization layer sets the `fuse` parameter as False and the default batchnorm sets it to True.
Im not sure how much of a problem this is, what do you think?

Also, I removed the `# wire` of `x` in the `build_resnet/effnet` functions. I think its better the function does 1 thing - create the model and the wiring can be done outside. It also better fits keras best practices ( I once saw a tweet about it from Chollet :) )

~~~python
in_ = tf.keras.layers.Input((224, 224, 3))
x = build_resnet()(in_)
tf.keras.Model(in_,x)
~~~

I wanted to discuss a bit before writing tests and fixing the notebooks and stuff...

### CODE REVIEW

File: tensorflow_similarity/architectures/efficientnet.py [hunk 1]
```diff
@@ -141,13 +144,10 @@ def EfficientNetSim(
     return SimilarityModel(inputs, outputs)
 
 
-def build_effnet(
-    x: layers.Layer, variant: str, weights: str, trainable: str
-) -> layers.Layer:
+def build_effnet(variant: str, weights: str = None, trainable: str = "full") -> tf.keras.Model:
```
Comment (owenvallis) [inline tensorflow_similarity/architectures/efficientnet.py:147]: We removed the `x: layers.Layer` here, so we'll need to update line 124 to:

`x = build_effnet(variant, weights, trainable)(x)`

File: tensorflow_similarity/architectures/efficientnet.py [hunk 2]
```diff
@@ -180,17 +181,12 @@ def build_effnet(
...
     else:
-        raise ValueError(
-            f"{trainable} is not a supported option for 'trainable'."
-        )
+        raise ValueError(f"{trainable} is not a supported option for 'trainable'.")
 
     # Don't train the BN layers if we are loading pre-trained weights.
     if weights:
         for layer in effnet.layers:
             if isinstance(layer, layers.BatchNormalization):
```
Comment (owenvallis) [inline tensorflow_similarity/architectures/efficientnet.py:189]: We'll need to change this to layers.experimental.SyncBatchNormalization?

File: tensorflow_similarity/architectures/resnet50.py [hunk 3]
```diff
@@ -103,12 +106,10 @@ def ResNet50Sim(
     return SimilarityModel(inputs, outputs)
 
 
-def build_resnet(x: layers.Layer, weights: str, trainable: str) -> layers.Layer:
+def build_resnet(weights: str = None, trainable: str = "full") -> tf.keras.Model:
```
Comment (owenvallis) [inline tensorflow_similarity/architectures/resnet50.py:109]: We removed the `x: layers.Layer` here, so we'll need to update line 86 to:

`x = build_resnet(variant, weights, trainable)(x)`

File: tensorflow_similarity/architectures/resnet18.py [hunk 4]
```diff
@@ -175,7 +175,7 @@ def block0(
         kernel_initializer=tf.keras.initializers.LecunUniform(),
         name=name + "_2_conv",
     )(x)
-    x = tf.keras.layers.BatchNormalization(
+    x = tf.keras.layers.experimental.SyncBatchNormalization(
```
Comment (owenvallis) [inline tensorflow_similarity/architectures/resnet18.py:178]: Since the other two build functions now return models, should we also have this return a model as well? It would be best if these were all consistent across the different architectures.

**Review by owenvallis — CHANGES_REQUESTED** [review]

**Comment by owenvallis** [comment]
Apologies for the delay in replying to the PR and thanks for adding support for this! The approach looks clean and I think the change to the build functions is cleaner. I left a few comments regarding breaking changes in the PR and I think we should have the resnet18 build function return a Model as well to be consistent with the other build functions. Otherwise looks good to me.

Once the comments are resolved, I'll run it on some of my benchmarking notebooks and verify the results are the same as before for the various losses.

**Comment by yonigottesman** [comment]
Thanks for the review. I was planning to fix all the stuff you pointed after you approve the general idea :) 
Fixed everything + tests.
Also, I rebased instead of the merge commit you added from main, It will make the repo history cleaner

**Review by owenvallis — APPROVED** [review]

**Comment by owenvallis** [comment]
Thanks again for the PR and for taking the extra time to clean up the git history. The changes look good to go.
